File: epi_judge_java/epi/SortedArrayRemoveDups.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TimedExecutor;

import java.util.List;

public class SortedArrayRemoveDups {
    // Returns the number of valid entries after deletion.
    public static int deleteDuplicates(List<Integer> list) {
        if (list.size() <= 1) return list.size();

        int uniqueIndex = 0;
        for (int i = 1; i < list.size(); i++) {
            if (!list.get(i).equals(list.get(uniqueIndex))) {
                list.set(++uniqueIndex, list.get(i));
            }
        }

        return uniqueIndex + 1;
    }

    @EpiTest(testDataFile = "sorted_array_remove_dups.tsv")
    public static List<Integer> deleteDuplicatesWrapper(TimedExecutor executor,
                                                        List<Integer> A)
            throws Exception {
        int end = executor.run(() -> deleteDuplicates(A));
        return A.subList(0, end);
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "SortedArrayRemoveDups.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/sorted_array_remove_dups.cc
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/timed_executor.h"

using std::vector;

// Returns the number of valid entries after deletion.
int DeleteDuplicates(vector<int> *A_ptr) {
    vector<int> &v = *A_ptr;
    if (v.empty()) return 0;

    int currIndex = 1;
    for (int i = 1; i < v.size(); ++i) {
        if (v[i] != v[i - 1]) {
            v[currIndex] = v[i];
            currIndex++;
        }
    }

    return currIndex;
}

vector<int> DeleteDuplicatesWrapper(TimedExecutor &executor, vector<int> A) {
    int end = executor.Run([&] { return DeleteDuplicates(&A); });
    A.resize(end);
    return A;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"executor", "A"};
    return GenericTestMain(
            args, "sorted_array_remove_dups.cc", "sorted_array_remove_dups.tsv",
            &DeleteDuplicatesWrapper, DefaultComparator{}, param_names);
}
